### PR TITLE
Add GTMGPC to Firedrake preconditioners

### PR DESCRIPTION
**Pull request by JDBetteridge**
I think this keeps meaning to be added to master and I'm not sure any additional changes are required.
This PR would add:

- Gopalakrishnan and Tan non-nested multigrid preconditioner `GTMGPC`
- A test case for `GTMGPC`
- Slightly modified `SCPC` to pass through a `dm` for the composed preconditioner
- Slightly modified `HybridisationPC` to pass through a `dm` for the composed preconditioner

This is largely Thomas' changes, but we are starting to build more on it.
Hence it would be beneficial to start the ball rolling on getting this into master, I've rebased this a second time to bring it inline with current master.

### CODE REVIEW

File: firedrake/preconditioners/gtmg.py [hunk 1]
```diff
@@ -0,0 +1,197 @@
...
+        pcmg.setType(pc.Type.MG)
+        pcmg.setOptionsPrefix(options_prefix)
+        pcmg.setMGLevels(2)
+        pcmg.setMGCycleType(pc.MGCycleType.V)
+        pcmg.setMGInterpolation(1, interp_petscmat)
+        pcmg.setOperators(A=fine_petscmat, P=fine_petscmat)
+
+        # We set a DM and an appropriate SNESContext for the coarse solver
+        # so we can do multigrid or patch solves.
+        from firedrake.variational_solver import NonlinearVariationalProblem
```
Comment (thomasgibson) [inline firedrake/preconditioners/gtmg.py]: Lines 143 -- 156 are definitely a hack. I think @wence- had a suggestion for improving this, but I have forgotten what was said. I think a utility method for `PCBase` which generates an appropriate SNES context for all subclassed PC objects would remove this hack (and prevent this snippet from being duplicated in `firedrake.SCPC` and `firedrake.HybridizationPC`).

File: firedrake/slate/static_condensation/hybridization.py [hunk 2]
```diff
@@ -228,14 +230,41 @@ def initialize(self, pc):
             nsp = nullspace(TraceSpace)
             Smat.setNullSpace(nsp.nullspace(comm=pc.comm))
 
-        # Set up the KSP for the system of Lagrange multipliers
+        # Create a SNESContext for the DM associated with the trace problem
```
Comment (thomasgibson) [inline firedrake/slate/static_condensation/hybridization.py]: See the previous comment in `preconditioners/firedrake/gtmg.py`

File: firedrake/slate/static_condensation/scpc.py [hunk 3]
```diff
@@ -119,17 +158,45 @@ def initialize(self, pc):
             nsp = nullspace(Vc)
             Smat.setNullSpace(nsp.nullspace(comm=pc.comm))
 
+        # Create a SNESContext for the DM associated with the condensed problem
```
Comment (thomasgibson) [inline firedrake/slate/static_condensation/scpc.py]: See the previous comment in `gtmg.py`

**Comment by thomasgibson** [comment]
Currently failing the pyadjoint test path. It looks unrelated to the changes made in this PR.

**Comment by thomasgibson** [comment]
Ok, now that #1609 is in, you can rebase this branch on top of master and remove the commits associated with #1609. Then I think it's good to go!

**Comment by wence-** [comment]
@JDBetteridge I cleaned this up and made sure it didn't differ from what had passed, so I think it's fine.

**Pull request merged**